vllm/entrypoints/openai/serving_embedding.py

```diff
@@ -133,6 +142,584 @@ def _build_response(
+                prompt_token_ids=chunk_tokens)
+
+            # Create chunk request prompt for logging
+            chunk_text = ""
+            chunk_request_prompt = TextTokensPrompt(
+                prompt=chunk_text, prompt_token_ids=chunk_tokens)
+
+            # Log the chunk
+            self._log_inputs(chunk_request_id,
+                             chunk_request_prompt,
+                             params=pooling_params,
+                             lora_request=ctx.lora_request,
+                             prompt_adapter_request=ctx.prompt_adapter_request)
+
+            # Create generator for this chunk
+            generator = self.engine_client.encode(
+                chunk_engine_prompt,
+                pooling_params,
+                chunk_request_id,
+                lora_request=ctx.lora_request,
+                trace_headers=trace_headers,
+                priority=getattr(ctx.request, "priority", 0),
+            )
+
+            generators.append(generator)
+
+        return generators
+
+    async def _aggregate_chunked_results(
+        self,
+        ctx: EmbeddingServeContext,
+        chunk_results: list[PoolingRequestOutput],
+        original_token_count: int,
+        original_prompt_token_ids: Optional[list[int]] = None,
+    ) -> PoolingRequestOutput:
+        """Aggregate results from multiple chunks using 
+        pooling-type-specific strategies."""
+        if len(chunk_results) == 1:
+            return chunk_results[0]
+
+        # Get pooling type to determine aggregation strategy
+        pooler_config = getattr(self.model_config, 'pooler_config', None)
+        pooling_type = getattr(pooler_config, 'pooling_type', 'MEAN')
+        if pooling_type:
+            pooling_type = pooling_type.upper()
+
+        # Route to appropriate aggregation method based on pooling type
+        if pooling_type in ['MEAN', 'AVG']:
+            return await self._aggregate_mean_pooling(
+                chunk_results, original_token_count, original_prompt_token_ids)
+        elif pooling_type == 'LAST':
+            return await self._aggregate_last_pooling(
+                chunk_results, original_prompt_token_ids)
+        elif pooling_type == 'CLS':
+            return await self._aggregate_cls_pooling(
+                chunk_results, original_prompt_token_ids)
+        else:
+            # For unsupported pooling types,
+            # fall back to mean aggregation with warning
+            logger.warning(
+                "Chunked aggregation for pooling type '%s' is not "
+                "specifically implemented. Falling back to weighted "
+                "averaging which may produce incorrect results.", pooling_type)
+            return await self._aggregate_mean_pooling(
+                chunk_results, original_token_count, original_prompt_token_ids)
+
+    async def _aggregate_mean_pooling(
+        self,
+        chunk_results: list[PoolingRequestOutput],
```

Comment:
By waiting to have the complete list of the chunks in CPU memory, you could potentially cause a large spike of memory usage. If there are many such requests, it could kill the server opening the door to denial of service attacks. There should be a configurable limit to the document size and the aggregation should be done online to keep the memory usage near-constant.